### PR TITLE
fix for itemsize => element_size() for torch backwards compat

### LINKED ISSUE #30304: AttributeError: 'torch.dtype' object has no attribute 'element_size'
### System Info

transformer version: 4.40.0.dev
python version: 3.10
torch: 2.0.1 cu11.7

I am fine-tuning https://huggingface.co/CohereForAI/c4ai-command-r-v01 model with axolotl framework. The config lora.yaml file is as follows:

```
base_model: CohereForAI/c4ai-command-r-v01
trust_remote_code: true

load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
    - path: Data_Clean3.json
      ds_type: json
      type: alpaca
dataset_prepared_path: /last_run_prepared/cohere-command/3308b18091e3a983103cbeb4cceb82d0
val_set_size: 0.0
output_dir: ./outputs/c4ai_lora

sequence_len: 1024
sample_packing: true
pad_to_sequence_len: true

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

adapter: qlora
lora_model_dir:
sample_packing: true
lora_r: 8
lora_alpha: 16
lora_dropout: 0.0
lora_target_modules:
  - q_proj
  - v_proj
  - k_proj
  - o_proj

gradient_accumulation_steps: 1
micro_batch_size: 1
num_epochs: 1
optimizer: paged_adamw_8bit
lr_scheduler: cosine
learning_rate: 0.0002

train_on_inputs: false
group_by_length: false
# bf16: auto
# fp16: 
# tf32: false
bf16: false
fp16: true
tf32: true

gradient_checkpointing: false  # don't use with fsdp_activation_checkpointing
gradient_checkpointing_kwargs:
  use_reentrant: false
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 0
saves_per_epoch: 1
debug:
weight_decay: 0.0
# deepspeed: deepspeed_configs/zero3.json

special_tokens:
  bos_token: "<BOS_TOKEN>"
  eos_token: "<|END_OF_TURN_TOKEN|>"
  pad_token: "<PAD>"
```

To reproduce, install the axolotl environment and run the following:
```
CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7" accelerate launch -m axolotl.cli.train examples/cohere-command/lora.yaml
```

The problem exists when I run the quantized model in 4 bit, in 8 bit it does not have any issues and run smoothly. 

### Who can help?

@pacman100 @SunMarc @younesbelkada 

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Install axolotl environment from: https://github.com/OpenAccess-AI-Collective/axolotl
2. run 
```
CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7" accelerate launch -m axolotl.cli.train examples/cohere-command/lora.yaml
```
3. You may change the dataset as mentioned in the axolotl environment.

### Expected behavior

It should start the training, but in 4_bit quantization qlora it's causing error.

### PR DESCRIPTION
**Pull request by winglian**
This is a similar fix to https://github.com/huggingface/peft/pull/1630.

`.itemsize` on a tensor is only supported on torch>=2.1

Fixes: https://github.com/huggingface/transformers/issues/30304

*Edit by @younesbelkada*:

This PR makes sure the checks are compatible with earlier versions of pytorch that have been overlooked by myself in https://github.com/huggingface/transformers/pull/30162

### CODE REVIEW

**Comment by winglian** [comment]
@pacman100 @BenjaminBossan

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1160,7 +1160,7 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
                 # used for the 4bit quantization (uint8 tensors are stored)
                 if is_loaded_in_4bit and isinstance(param, bnb.nn.Params4bit):
                     total_numel.append(
-                        param.numel() * 2 * self.hf_quantizer.quantization_config.bnb_4bit_quant_storage.itemsize
+                        param.numel() * 2 * self.hf_quantizer.quantization_config.bnb_4bit_quant_storage.element_size()
```
Comment (muellerzr) [inline src/transformers/modeling_utils.py:1163]: I meant to respond to the other one earlier, but may be good to have an if/else to know which version to call? E.g. `param.numel() * 2 * elem_size` where `elemsize` is defined earlier based on the torch version

**Comment by BenjaminBossan** [comment]
Note that this PR [fails](https://github.com/huggingface/peft/pull/1630#issuecomment-2044440626) on PEFT CI.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/modeling_utils.py:1163]: Correction, as @BenjaminBossan pointed out this is a must for us to have

**Comment by winglian** [comment]
my last change isn't correct either. I'll take a look at this tomorrow.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (peterdonnelly1) [reply src/transformers/modeling_utils.py]: Besides itemsize not being available in torch 2.0.1, I couldn't get element_size() to work either (not a dtype attribute error).   

I put this 'helper' function at the top of the file to do it manually:

def get_dtype_size(dtype):
    if dtype == torch.float32:
        return 4
    elif dtype == torch.float64:
        return 8
    elif dtype == torch.float16:
        return 2
    elif dtype == torch.uint8:
        return 1
    elif dtype == torch.int8:
        return 1
    elif dtype == torch.int16:
        return 2
    elif dtype == torch.int32:
        return 4
    elif dtype == torch.int64:
        return 8
    else:
        raise ValueError("Unsupported dtype")

and then used this in place of the itemsize code:

if is_loaded_in_4bit and isinstance(param, bnb.nn.Params4bit):
    quant_storage = self.hf_quantizer.quantization_config.bnb_4bit_quant_storage
    nb_params = get_dtype_size(quant_storage)
    total_numel.append(param.numel() * 2 * nb_params)
else:
    total_numel.append(param.numel())
    
I've only tested it by using the qlora config on tiny-llama; and at least for this case it works:

   accelerate launch -m axolotl.cli.train examples/tiny-llama/qlora.yml

I am using image 'winglian/axolotl-base:main-base-py3.11-cu121-2.2.1' (which actually uses torch 2.0.1 not 2.2.1 as the name implies.).  Ubuntu 22.04 / laptop RTX4090 (16G)

**Comment by younesbelkada** [comment]
Hi @winglian , as pointed out by @hiyouga my fix did not cover all the cases, would you be happy to rebase your PR with main and we merge it ?

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1160,12 +1160,15 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
                 # For 4bit models, we need to multiply the number of parameters by 2 as half of the parameters are
                 # used for the 4bit quantization (uint8 tensors are stored)
                 if is_loaded_in_4bit and isinstance(param, bnb.nn.Params4bit):
-                    quant_storage = self.hf_quantizer.quantization_config.bnb_4bit_quant_storage
-                    # For compatibility with older PT version - see: https://github.com/huggingface/peft/pull/1635
-                    nb_params = (
-                        quant_storage.itemsize if hasattr(quant_storage, "itemsize") else quant_storage.element_size()
+                    if hasattr(param, "element_size"):
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:1163]: Why not simply use:

```py
quant_storage = self.hf_quantizer.quantization_config.bnb_4bit_quant_storage
num_bytes = quant_storage.element_size()
```

`element_size` seems present from 1.9.1: https://pytorch.org/docs/1.9.1/search.html?q=element_size&check_keywords=yes&area=default to latest: https://pytorch.org/docs/2.2/search.html?q=element_size&check_keywords=yes&area=default

Comment (younesbelkada) [reply src/transformers/modeling_utils.py:1163]: cc @winglian @BenjaminBossan @hiyouga

Comment (hiyouga) [reply src/transformers/modeling_utils.py:1163]: `self.hf_quantizer.quantization_config.bnb_4bit_quant_storage` is a `torch.dtype` instance while only the `torch.tensor` has `element_size()`

**Review by younesbelkada — APPROVED** [review]
Thanks for taking care of the backward compatibility for previous torch versions !

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -1160,12 +1160,13 @@ def num_parameters(self, only_trainable: bool = False, exclude_embeddings: bool
...
-                    nb_params = (
-                        quant_storage.itemsize if hasattr(quant_storage, "itemsize") else quant_storage.element_size()
-                    )
-                    total_numel.append(param.numel() * 2 * nb_params)
+                    if hasattr(param, "element_size"):
+                        num_bytes = param.element_size()
+                    elif not hasattr(param, "quant_storage"):
+                        num_bytes = 1
+                    else:
+                        num_bytes = param.quant_storage.itemsize
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:1168]: nit - it's clearer to reference the attr `quant_storage` on the logic check that checks whether or not the object has this attribute 

```suggestion
                    elif hasattr(param, "quant_storage"):
                        num_bytes = param.quant_storage.itemsize
                    else:
                        num_bytes = 1
```

**Review by amyeroberts — APPROVED** [review]
Thanks for handling this fix! 

Just a small nit

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (yand1ngs) [reply src/transformers/modeling_utils.py]: Thank you soooo much!!!

**Pull request merged**